### PR TITLE
Prometheus query builder: Replace select with AsyncSelect to support high cardinality prometheus instances

### PR DESCRIPTION
**Pull request by gtk-grafana**
**What this PR does / why we need it**:
In order to resolve reported issues with high cardinality prometheus instances we need a better short-term solution to manage pulling metric names into the query builder. There is a discussion a long term solution in [this epic](https://github.com/grafana/observability-metrics-squad/issues/47)

This PR attemps to do two things:
1. Limit the number of results we display in selects to 1000
2. Utilizes the backend API to facilitate autocomplete functionality instead of client-side search within the results currently saved to the state of the select element.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/support-escalations/issues/4008

**Special notes for your reviewer**:
* String manipulation like this is kinda hard to debug, is there a cleaner implementation we can pursue?
* The clientside search didn't get removed, it's just limited to the results that are already displayed in the dropdown, I attached a video that demonstrates this functionality, but basically the clientside search will be applied while the API call to the prometheus API is going on, so for users with less than 1000 results, they'll see instant results (although we'll still attempt to load more results from the backend).
* I attempted to reproduce all the functionality of the clientside search in the call to the prometheus API, however I could have missed stuff if any functionality of the client side search isn't covered by the unit tests

### CODE REVIEW

**Comment by gtk-grafana** [comment]
Here's a quick demo:
https://user-images.githubusercontent.com/109082771/196777885-0c0c8188-50b4-44eb-836a-bb50d76fd63c.mov

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx [hunk 1]
```diff
@@ -56,25 +63,71 @@ export function MetricSelect({ query, onChange, onGetMetrics }: Props) {
...
+  };
+
+  const debouncedSearch = debounce((query: string) => getMetricLabels(query), 300);
+
   return (
     <EditorFieldGroup>
       <EditorField label="Metric">
-        <Select
+        <AsyncSelect
+          width={'auto'}
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx:112]: this is default, i think you can remove

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx [hunk 2]
```diff
@@ -56,25 +63,87 @@ export function MetricSelect({ query, onChange, onGetMetrics }: Props) {
...
+  const getMetricLabels = (query: string) => {
+    // Since some customers can have millions of metrics, whenever the user changes the autocomplete text we want to call the backend and request all metrics that match the current query string
+    const results = datasource.metricFindQuery(formatKeyValueStringsForLabelValuesQuery(query, labelsFilters));
+    return results.then((results) => {
+      if (results.length > MAX_NUMBER_OF_RESULTS) {
+        results.splice(0, results.length - MAX_NUMBER_OF_RESULTS);
+      }
+      return results.map((result) => {
+        return {
+          label: result.text,
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx:116]: are both label and value required? Can we cut down the size of this object by only passing in one?

Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/querybuilder/components/MetricSelect.tsx:116]: Nope, both required

**Review by itsmylife — APPROVED** [review]

**Pull request merged**